File: leetcode/5_LongestPS/lps.go
package lps

func longestPalindrome(s string) string {
	if len(s) == 0 {
		return ""
	}
	i1, i2 := 0, len(s)
	start, end := 0, 1
	for i1 < i2 {
		if isPalindrome(s[i1:i2]) {
			if (i2 - i1) > (end - start) {
				start, end = i1, i2
			}
		}
		i2--
		if i2 == i1 {
			i1++
			i2 = len(s)
			if (i2 - i1) < (end - start) {
				break
			}
		}
	}
	return s[start:end]
}

func isPalindrome(s string) bool {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		if s[i] != s[j] {
			return false
		}
	}
	return true
}

func longestPalindrome2(s string) string {
	if len(s) == 0 {
		return ""
	}
	start,end := 0,0
	for i:=0;i<len(s);i++ {
		l1 :=  expandAroundCenter(s,i,i)
		l2 := expandAroundCenter(s,i,i+1)
		ll := max(l1,l2)
		if ll > (end - start){
			start = i - (ll -1)/2
			end  = i + ll /2
		}
	}
	return s[start:end+1]
}

func expandAroundCenter(s string, l,r int) int {
	for  ;l >= 0 && r < len(s) && s[l] == s[r]; {
		l--
		r++
	}
	return r - l - 1;
}

func max(a,b int) int {
	if a > b {
		return a
	}
	return b
}

File: leetcode/5_LongestPS/lps.rs
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        if s.len() == 0 {
            return ""
        }
        let mut start: usize  =  0;
        let mut end: usize = 0;

        for i in (0..s.len()) {
            let l1 = expand_around_center(&s, i, i);
            let l2 = expand_around_center(&s, i, i+1);
            let ll = l1.max(l2);
            if ll > (end - start) {
                start = i - (ll - 1)/2;
                end = i + ll / 2;
            }
        }
        return String::from(s[start..end+1])
    }
    fn expand_around_center(s: &str, mut l: usize,mut r: usize) -> usize {
        while l >= 0 && r < s.len() && s[l] == s[r] {
            l-=1;
            r+=1;
        }
        return r-l-1
    }
}